docs/rustdoc/prehashing.md

```diff
@@ -0,0 +1,183 @@
+- If $n_p$ is too small, bins may not be sufficiently populated, reducing the chances of constructing valid tuples.
+
+### Example walkthrough
+- **Prover’s set** $S_p = (A, B, C, D, E)$
+- **Proof length** $u = 3$
+- **Search width** $d = 2$ (i.e., $t \in \{1, 2\}$)
+---
+> **Step 1:** Prehash the elements with $H_0$.
+> - The prover precomputes $H_0(s)$ for all elements in $S_p$, assigning each element to a bin:
+>   - $H_0(A) = 1$
+>   - $H_0(B) = 2$
+>   - $H_0(C) = 1$
+>   - $H_0(D) = 4$
+>   - $H_0(E) = 4$
+> - Bins are grouped as follows:
+>   - $\mathsf{bin_1}$: $\[A, C\]$
+>   - $\mathsf{bin_2}$: $\[B\]$
+>   - $\mathsf{bin_3}$: $\emptyset$
+>   - $\mathsf{bin_4}$: $\[D, E\]$
+> ---
+> **Step 2:** Proof construction for $t = 1$.
+> - The prover starts with $t = 1$ and attempts to construct a valid sequence of length $u = 3$.
+> - Depth $k = 1$:
+>   - Compute $H_1(1)$ to determine the first bin to explore.
+>     - Assume $H_1(1) = 1$, so the prover looks in bin 1.
+>   - $\mathsf{bin_1}$: $\[A, C\]$
+>     - Extend the sequence with $s_1 = A$.
+>   - Current sequence: $(t, s_1) = (1, A)$.
+> - Depth $k = 2$:
+>   - Compute $H_1(t, s_1)$ to determine the next bin to explore.
+>     - Assume $H_1(1, A) = 3$, so the prover looks in bin 3.
+>   - $\mathsf{bin_3}$: $\emptyset$
+>     - No valid $s_2$ can be selected.
+> - Backtracking for $t = 1$:
+>   - After exhausting all possible sequences for $t = 1$, no valid proof is found.
+>   - The prover backtracks and tries $t = 2$.
+> ---
+> **Step 3:** Proof construction for $t = 2$.
+> - Depth $k = 1$:
+>   - Compute $H_1(2)$ to determine the first bin to explore.
+>     - Assume $H_1(2) = 4$, so the prover looks in bin 4.
+>   - $\mathsf{bin_4}$: $\[D, E\]$
+>     - Extend the sequence with $s_1 = D$.
+>   - Current sequence: $(t, s_1) = (2, D)$.
+> - Depth $k = 2$:
+>   - Compute $H_1(t, s_1)$ to determine the next bin to explore.
+>     - Assume $H_1(2, D) = 2$, so the prover looks in bin 2.
+>   - $\mathsf{bin_2}$: $\[B\]$
+>     - Extend the sequence with $s_2 = B$.
+>   - Current sequence: $(t, s_1, s_2) = (2, D, B)$.
+> - Depth $k = 3$:
+>   - Compute $H_1(2, s_1, s_2)$ to determine the next bin to explore.
+>     - Assume $H_1(2, D, B) = 1$, so the prover looks in bin 1.
+>   - $\mathsf{bin_1}$: $\[A, C\]$
+>     - Extend the sequence with $s_3 = C$.
+>   - Current sequence: $(t, s_1, s_3, s_3) = (2, D, B, C)$.
+> ---
+> **Step 4:** Final Validation.
+> - A tuple with size $u = 3$ is obtained: $(t, s_1, s_2, s_3) = (2, D, B, C)$.
+> - If $H_2(2, D, B, C) = 1$, the sequence is valid, and the prover outputs it as proof.
+---
+
+## Functions
+#### Proving
+---
+> $\mathsf{Prove}(S_p) \rightarrow \pi$
+> - Input:
+>   - $S_p:~~$ `prover_set`, a set of elements to be proven.
+> - Output:
+>   - $\pi:~~$ `proof`, a valid proof $(t, s_1, \ldots, s_u)$ or $\bot$.
+> ---
+> - $bins \leftarrow$ { }
+> - **for** each $~~ i \in \[1,~ n_p\]:$
```

Comment:
This is a bit confusing. Here `[ ]` denotes an interval, but above for bins it denoted a list or a set. Suggest to use for bins either `( )` if it's a list or `{ }` if it's a set.